src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp

```diff
@@ -992,6 +993,233 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_JoinerStart(voi
+        break;
+    case OT_ERROR_DUPLICATED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Domain name or RRset is duplicated");
+        break;
+    case OT_ERROR_RESPONSE_TIMEOUT:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Timed out waiting on server response");
+        break;
+    case OT_ERROR_INVALID_ARGS:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Invalid service structure detected");
+        break;
+    case OT_ERROR_NO_BUFS:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Insufficient buffer to handle message");
+        break;
+    case OT_ERROR_FAILED:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Internal server error occurred");
+        break;
+    default:
+        ChipLogError(DeviceLayer, "OnSrpClientNotification: Unknown error occurred");
+        break;
+    }
+}
+
+template <class ImplClass>
+void GenericThreadStackManagerImpl_OpenThread<ImplClass>::_InitSrpClient()
+{
+    otSrpClientSetCallback(mOTInst, &OnSrpClientNotification, nullptr);
+}
+
+template <class ImplClass>
+CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_AddSrpService(const char * aInstanceName, const char * aName,
+                                                                               uint16_t aPort, uint32_t aLeaseInterval,
+                                                                               uint32_t aKeyLeaseInterval)
+{
+    otError error                   = OT_ERROR_NONE;
+    otSrpClientService * srpService = nullptr;
+
+    VerifyOrExit(aInstanceName, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aName, error = OT_ERROR_INVALID_ARGS);
+
+    // Check if service with desired instance name already exists.
+    for (otSrpClientService & service : mSrpServices)
+    {
+        if (service.mInstanceName)
+        {
+            VerifyOrExit((strcmp(service.mInstanceName, aInstanceName) != 0), error = OT_ERROR_DUPLICATED);
+        }
+        else
+        {
+            srpService = srpService ? srpService : &service;
+        }
+    }
+
+    // Verify is there an empty place for new service.
+    VerifyOrExit(srpService, error = OT_ERROR_NO_BUFS);
+
+    memset(srpService, 0, sizeof(otSrpClientService));
+    otSrpClientSetLeaseInterval(mOTInst, aLeaseInterval);
+    otSrpClientSetKeyLeaseInterval(mOTInst, aKeyLeaseInterval);
+
+    srpService->mInstanceName = aInstanceName;
+    srpService->mName         = aName;
+    srpService->mPort         = aPort;
+
+    error = otSrpClientAddService(mOTInst, srpService);
+
+exit:
+    return MapOpenThreadError(error);
+}
+
+template <class ImplClass>
+CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::_RemoveSrpService(const char * aInstanceName, const char * aName)
+{
+    otError error                   = OT_ERROR_NONE;
+    otSrpClientService * srpService = nullptr;
+
+    VerifyOrExit(aInstanceName, error = OT_ERROR_INVALID_ARGS);
```

Comment:
Wouldn't it be confusing to see OT errors in case it's actually an issue in CHIP?

Reply:
Done